File: golang-devildyw/hard/L23合并K个升序链表.go
package main

func main() {}

func mergeKLists(lists []*ListNode) *ListNode {
	n := len(lists)
	if n <= 0 {
		return nil
	}

	for n > 1 {
		index := 0
		for i := 0; i < n; i = i + 2 {
			if i+2 > n {
				lists[index] = lists[i]
			} else {
				newHead := merge(lists[i], lists[i+1])
				lists[index] = newHead
			}
			index++
		}
		n = index
	}
	return lists[0]
}

func merge(head1 *ListNode, head2 *ListNode) *ListNode {
	dummyNode := &ListNode{}
	cur := dummyNode
	for head1 != nil && head2 != nil {
		if head1.Val <= head2.Val {
			cur.Next = head1
			head1 = head1.Next
		} else {
			cur.Next = head2
			head2 = head2.Next
		}
		cur = cur.Next
	}

	if head1 != nil {
		cur.Next = head1
	}
	if head2 != nil {
		cur.Next = head2
	}

	return dummyNode.Next
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L23合并K个升序链表.java
package top.dyw.leetcode.Hot100;

public class L23合并K个升序链表 {
    public ListNode mergeKLists(ListNode[] lists) {
        int n = lists.length;
        if (n==0) {
            return null;
        }

        int subLength = 1;
        while (subLength < n) {
            for (int i=0; i<n; i+=2*subLength) {
                if (i+subLength >= n) {
                    break;
                }
                ListNode head1 = lists[i];
                ListNode head2 = lists[i+subLength];

                lists[i] = mergeList(head1,head2);
            }

            subLength <<=1;
        }

        return lists[0];
    }

    public ListNode mergeList(ListNode head1, ListNode head2) {
        ListNode dummyNode = new ListNode(-1);

        ListNode cur = dummyNode;
        while (head1!=null && head2!=null) {
            if (head1.val < head2.val) {
                cur.next= new ListNode(head1.val);
                head1 = head1.next;
            } else {
                cur.next = new ListNode(head2.val);
                head2 = head2.next;
            }

            cur = cur.next;
        }

        if (head1!=null) {
            cur.next = head1;
        }

        if (head2!=null) {
            cur.next = head2;
        }
        return dummyNode.next;
    }
}
